opnode/rollup/driver/state.go

```diff
@@ -86,51 +114,160 @@ func (s *state) l1WindowBufEnd() eth.BlockID {
+// if the next L2 block's time is greater than or equal to the L2 Head's Origin.
 func (s *state) findNextL1Origin(ctx context.Context) (eth.L1BlockRef, error) {
-	// [prev L2 + blocktime, L1 Bock)
+	// If we are at the head block, don't do a lookup.
+	// Don't do a timestamp check either as we are unable to get the next block even if we wanted to.
 	if s.l2Head.L1Origin.Hash == s.l1Head.Hash {
 		return s.l1Head, nil
 	}
+
+	// Grab the block ref
 	curr, err := s.l1.L1BlockRefByHash(ctx, s.l2Head.L1Origin.Hash)
 	if err != nil {
 		return eth.L1BlockRef{}, err
 	}
+	// Somehow reorg'd. Will let the state loop take care of it.
+	if curr.Hash != s.l2Head.L1Origin.Hash {
+		return eth.L1BlockRef{}, errors.New("Unknown L1Origin")
+	}
 
-	s.log.Info("Find next l1Origin", "l2Head", s.l2Head, "l1Origin", curr)
 	// TODO: There is an interaction with not using the L1 Genesis as an L1 Origin and
 	// the fact that the L2 Genesis time needs to be set around the L1 Genesis such
 	// that this check will return true.
 	if s.l2Head.Time+s.Config.BlockTime >= curr.Time {
 		// TODO: Need to walk more?
 		ref, err := s.l1.L1BlockRefByNumber(ctx, curr.Number+1)
-		s.log.Info("Looking up new L1 Origin", "nextL1Origin", ref)
+		s.log.Debug("Advancing L1 Origin", "l2Head", s.l2Head, "previous_l1Origin", s.l2Head.L1Origin, "l1Origin", ref, "err", err)
 		return ref, err
 	}
+	s.log.Debug("Next L1 Origin is the same as the previous", "l2Head", s.l2Head, "l1Origin", curr)
 	return curr, nil
 }
 
+// createNewL2Block builds a L2 block on top of the L2 Head (unsafe)
+func (s *state) createNewL2Block(ctx context.Context) (eth.L1BlockRef, error) {
+	nextOrigin, err := s.findNextL1Origin(context.Background())
+	if err != nil {
+		s.log.Error("Error finding next L1 Origin", "err", err)
+		return eth.L1BlockRef{}, err
+	}
+	if nextOrigin.Time <= s.Config.BlockTime+s.l2Head.Time {
+		s.log.Trace("Skipping block production because the next block time is behind the next L1 Origin", "l2Head", s.l2Head, "l1Origin", nextOrigin)
+		return eth.L1BlockRef{}, nil
+	}
+	// Don't produce blocks until past the L1 genesis
+	if nextOrigin.Number <= s.Config.Genesis.L1.Number {
+		s.log.Trace("Skipping block production because the next L1 Origin is behind the L1 genesis")
+		return eth.L1BlockRef{}, nil
+	}
+	// Actually create the new block
+	newUnsafeL2Head, batch, err := s.output.createNewBlock(context.Background(), s.l2Head, s.l2SafeHead.ID(), s.l2Finalized, nextOrigin.ID())
+	if err != nil {
+		s.log.Error("Could not extend chain as sequencer", "err", err, "l2UnsafeHead", s.l2Head, "l1Origin", nextOrigin)
+		return eth.L1BlockRef{}, err
+	}
+	// State update
+	s.l2Head = newUnsafeL2Head
+	s.log.Info("Sequenced new l2 block", "l2Head", s.l2Head, "l1Origin", s.l2Head.L1Origin)
+	//Submit batch
+	go func() {
+		_, err := s.bss.Submit(&s.Config, []*derive.BatchData{batch}) // TODO: submit multiple batches
+		if err != nil {
+			s.log.Error("Error submitting batch", "err", err)
```

Comment:
This log is unsafe and will make the thing panic if the logger is closed before the go routine runs. (happened in older version during testing once for me). It's going to change with the new batch-submitter though. Maybe just add a comment for the audit that this batch-submission is going to change with #312 and that issue #308 is open to improve the life cycle of the node?

Reply:
Noted issue 308